src/mgr/DaemonServer.cc

```diff
@@ -1643,20 +1645,26 @@ bool DaemonServer::_handle_command(
 	    continue;
 	  }
 	  touched_pgs++;
+
+	  const pg_pool_t *pi = osdmap.get_pg_pool(q.first.pool());
+	  if (!pi) {
+	    ++dangerous_pgs; // pool is creating or deleting
+            continue;
+	  }
+
 	  if (!(q.second.state & PG_STATE_ACTIVE) ||
-	      (q.second.state & PG_STATE_DEGRADED)) {
+              ((pi->type != pg_pool_t::TYPE_ERASURE) &&
+               (q.second.state & PG_STATE_DEGRADED))) {
 	    ++dangerous_pgs;
 	    continue;
 	  }
-	  const pg_pool_t *pi = osdmap.get_pg_pool(q.first.pool());
-	  if (!pi) {
-	    ++dangerous_pgs; // pool is creating or deleting
-	  } else {
-	    if (pg_acting.size() < pi->min_size) {
+	    if (((q.second.state & PG_STATE_DEGRADED) &&
+                  !avail_no_missing.size() &&
+                  (pi->type == pg_pool_t::TYPE_ERASURE)) ||
```

Comment:
should be pi->type != pg_pool_t::TYPE_ERASURE

Reply:
You could remove _avail_no_missing_ set and just:

if (((q.second.state & PG_STATE_DEGRADED) &&
    !pg_acting.size() &&
    (pi->type != pg_pool_t::TYPE_ERASURE)) ||

Reply:
I still think you aren't changing anything with this change.  If _pg_acting.size() == 0_ then it will be less than _min_size_.  Remember for degraded _avail_no_missing == pg_acting_ which is also why you don't need the _avail_no_missing_ set.

Reply:
> When a pg is degraded the _pg_acting_ is based on _avail_no_missing_. If not, pg_acting comes from _acting_. In any case the size of _pg_acting_ must be >= _min_size_. Unfound is for objects that aren't stored on any OSD.


Yes, but _pg_acting.size() >= min_size_ doesn't necessarily mean _avail_no_missing_ is not empty, as it looks to me that pg_acting comes from _acting_ and avail_no_missing is constructed when PeeringState do "update_calc_stats". What I wanted to express here is that "if after stopping the specified osds would mean that there's no OSD who has all the up-to-date objects of the pg is still up in the cluster, it's not ok to stop even if the pg can go into PG_STATE_ACTIVE".
 
> Now that think about it, I'm not clear how a pg with pg_acting.size() == min_size wouldn't be dangerous because stopping an OSD could then cause a PG to drop below min_size. Doesn't "available" imply writable.

Yes, but if, say, a pool's size is 5 and min_size is 2, the current implementation doesn't allow stopping anymore OSDs, even if currently there's only one OSD is down. Actually, according to our online cluster's "ops" team, cases like this happens. So we think maybe the current ok-to-stop is a little bit rigorous. If we want to avoid pg_acting.size() == min_size, we can modify the code to something like:
```
if (pg_acting.size() <= pi->min_size)
    ++dangerous_pgs;
```

Am I right?


> I still think you aren't changing anything with this change. If _pg_acting.size() == 0_ then it will be less than _min_size_. Remember for degraded _avail_no_missing == pg_acting_ which is also why you don't need the _avail_no_missing_ set.

Um...it looks to me that this is not the case. avail_no_missing may be less than pg_acting.

Reply:
> should be pi->type != pg_pool_t::TYPE_ERASURE

Sorry, this must be a typo......will change it.

Reply:
> Yes, but if, say, a pool's size is 5 and min_size is 2, the current implementation doesn't allow stopping anymore OSDs, even if currently there's only one OSD is down. Actually, according to our online clusters' "ops" team, cases like this happens. So we think maybe the current ok-to-stop is a little bit rigorous. If we want to avoid pg_acting.size() == min_size, we can modify the code to something like:

If there is already an OSD down in your example, then pg_acting might be size 4 which is 4 < 2 (min_size) is NOT dangerous as expected. I don't see why that would or should prevent another OSD from being stopped.

> 
> ```
> if (pg_acting.size() <= pi->min_size)
>     ++dangerous_pgs;
> ```
> 
> Am I right?

Yes, I'm not sure why the code doesn't already have <=.  Of course, that is more stringent.

> 
> > I still think you aren't changing anything with this change. If _pg_acting.size() == 0_ then it will be less than _min_size_. Remember for degraded _avail_no_missing == pg_acting_ which is also why you don't need the _avail_no_missing_ set.
> 
> Um...it looks to me that this is not the case. avail_no_missing may be less than pg_acting.

See my other comment.  Don't confuse the added local _avail_no_missing_ and _q.second.avail_no_missing_ nor _pg_acting_ and _q.second.acting_.

Reply:
> > Yes, but if, say, a pool's size is 5 and min_size is 2, the current implementation doesn't allow stopping anymore OSDs, even if currently there's only one OSD is down. Actually, according to our online clusters' "ops" team, cases like this happens. So we think maybe the current ok-to-stop is a little bit rigorous. If we want to avoid pg_acting.size() == min_size, we can modify the code to something like:
> 
> If there is already an OSD down in your example, then pg_acting might be size 4 which is 4 < 2 (min_size) is NOT dangerous as expected. I don't see why that would or should prevent another OSD from being stopped.

Oh, sorry that I made this misunderstanding. What is preventing OSD to stop is not this line, but [here](https://github.com/ceph/ceph/blob/718af4faaf47c416544f872d2932fce52e60549b/src/mgr/DaemonServer.cc#L1647)

> 
> > ```
> > if (pg_acting.size() <= pi->min_size)
> >     ++dangerous_pgs;
> > ```
> > 
> > 
> > Am I right?
> 
> Yes, I'm not sure why the code doesn't already have <=. Of course, that is more stringent.
> 
> > > I still think you aren't changing anything with this change. If _pg_acting.size() == 0_ then it will be less than _min_size_. Remember for degraded _avail_no_missing == pg_acting_ which is also why you don't need the _avail_no_missing_ set.
> > 
> > 
> > Um...it looks to me that this is not the case. avail_no_missing may be less than pg_acting.
> 
> See my other comment. Don't confuse the added local _avail_no_missing_ and _q.second.avail_no_missing_ nor _pg_acting_ and _q.second.acting_.

Finally, I got your point. Will remove avail_no_missing set:-)